src/cheri_insts.sail

```diff
@@ -1370,6 +1370,66 @@ function clause execute (LoadCapCap(cd, cs1)) = {
   handle_load_cap_via_cap(cd, 0b0 @ cs1, cs1_val, vaddr)
 }
 
+union clause ast = CLoadTags : (regidx, regidx)
+function clause execute (CLoadTags(rd, cs1)) = {
+  let cs1_val = C(cs1);
+  if not (cs1_val.tag) then {
+    handle_cheri_reg_exception(CapEx_TagViolation, cs1);
+    RETIRE_FAIL
+  } else if isCapSealed(cs1_val) then {
+    handle_cheri_reg_exception(CapEx_SealViolation, cs1);
+    RETIRE_FAIL
+  } else if not (cs1_val.permit_load) then {
+    handle_cheri_reg_exception(CapEx_PermitLoadViolation, cs1);
+    RETIRE_FAIL
+  } else if not (cs1_val.permit_load_cap) then {
+    handle_cheri_reg_exception(CapEx_PermitLoadCapViolation, cs1);
+    RETIRE_FAIL
+  } else {
+    let vaddr = cs1_val.address;
+    if not (inCapBounds(cs1_val, vaddr, caps_per_cacheline * cap_size)) then {
+      handle_cheri_reg_exception(CapEx_LengthViolation, cs1);
+      RETIRE_FAIL
+    } else if not (unsigned(vaddr) % (caps_per_cacheline * cap_size) == 0) then {
+      handle_mem_exception(vaddr, E_Load_Addr_Align());
+      RETIRE_FAIL
+    } else match translateAddr(vaddr, Read(Cap)) {
+      TR_Failure(e, _) => { handle_mem_exception(vaddr, e); RETIRE_FAIL },
+      TR_Address(addr, ptw_info) => {
+	match ptw_info.ptw_lc {
+          PTW_LC_CLEAR => {
+            handle_mem_exception(vaddr, E_Extension(EXC_LOAD_CAP_PAGE_FAULT));
+            RETIRE_FAIL
+          },
+          PTW_LC_TRAP => {
+            handle_mem_exception(vaddr, E_Extension(EXC_LOAD_CAP_PAGE_FAULT));
+            RETIRE_FAIL
+          },
+          PTW_LC_OK => {
+            x : MemoryOpResult(xlenbits) = MemValue(zeros());
+
+            foreach (i from 0 to (caps_per_cacheline - 1)) {
+              let c = mem_read_cap(addr, false, false, false);
+              match (x,c) {
+		(MemException(e), _) => { () },
+                (MemValue(xv), MemValue(cv)) => {
+                  y = xv; y[i] = bool_to_bits(cv.tag)[0]; x = MemValue(y)
+                },
+                (_, MemException(e)) => { x = MemException(e) }
+              }
+            };
```

Comment:
Do we really not have bind or an imperative equivalent? We should have one and remove this gunk... I guess in this case what we really want is a `sequence :: Monad m => [m a] -> m [a]` equivalent; could put one in cheri_mem.sail and then eventually move it into riscv_mem.sail. Then you could write:
```sail
res : vector(caps_per_cacheline, dec, MemoryOpResult(bit));
foreach (i from 0 to caps_per_cacheline)
  res[i] = mem_read_cap(addr, false, false, false);

match MemoryOpResult_sequence(res) {
  ...
}
```

I'll have a play with that :)

Reply:
I confess my familiarity with sail's base library is... lacking.  If there's a `sequence` laying around, I'm not aware of it.  If there's also a `bind`, we could simplify some of the PTW logic the Revocation 3 PTE bits have introduced.

Reply:
(sequence didn't make it much better)